Firmware Signing Post (#136)
memfault/interrupt: _drafts/fwup-signing.md

---
title: "Authenticate firmware builds with code signing"
description: WIP
author: francois
---

<!-- excerpt start -->
<!-- excerpt end -->

{% include newsletter.html %}
{% include toc.html %}

## Firmware Signing Explained

### What is code signing

Code signing is a method of proving a file was created by a trusted source and
has not been tampered with. This is achieved by creating a signature for the
file: a token which can be verified but not forged.

We all use signatures in our day to day life, by scribbling our name at the
bottom of our credit card receipts, contracts, and bank checks. These signatures
can be verified quite easily by comparing them to previous signatures, but are
hard to forge because they rely on an individual's muscle memory.

Code signing uses cryptographic algorithms to achieve a similar goal. The
resulting signatures can only be generated by people who know a given password,
also known as a secret key, but can be verified as genuine by anyone.

Additionally, the signatures are tied to a given firmware binary so should the

> binary change by so much as 1 bit the signature would no longer be valid.

Reviewer: I'm struggling with words today, but you jump right into firmware binaries without yet talking about firmware binaries. I think I'm missing how this all applies to firmware from a high level.

Something like:

When a signature is generated for a firmware binary, it is uniquely tied to that exact firmware. Changing the binary so much as 1 bit would produce a different signature.

With this in mind, if a device is given a signed binary and a signature and can verify both that the binary and signature are valid and that the binary was created by the correct author, then we know the binary has not been tampered with.

Basically, connect the dots for me, either in a diagram or words, before we get into this stuff where we are dealing directly with openssl

Author: Used your words, they're great

### Why sign our firmware

By implementing signature verification in our bootloader we can identify
whether or not a given firmware update was provided by the manufacturer, or if
it has been tampered with. The bootloader can then decide to either warn the
user, void the device's warranty, or simply refuse to run the unauthenticated
binary.

> With more and more devices connected to the internet, security is an

Reviewer: ```suggestion
With more and more devices connected to the Internet, security is an
```

increasingly hot topic in firmware development. A device which accepts firmware

> updates over the wireless or internet connectivity but does not verify it opens

Reviewer: ```suggestion
updates wirelessly or over Internet connectivity but does not verify it opens
```

itself to compromise. By feeding it with a malicious firmware image, an attacker
might:
* Brick the device, or the whole fleet
* Snoop on end users and compromise their privacy and security

> * Strategically malfunction at a critical time

Reviewer: If you want to open the router can of worms...create a botnet of Linux routers for funsies.

Author: I decided not to call anyone out explicitly.

These are highly undesirable outcomes, which can be effected at scale due to the

> internet of things. In 2020, it is reckless to implement firmware update for our

Reviewer: ```suggestion
Internet of things. In 2020, it is reckless to implement firmware update for our
```

systems without some form of authentication.

> > **What signing is not**: code signing is an important component of firmware

Reviewer: Maybe even move it to its own section rather than a big block comment.

Author: I don't think I have quite enough for a section

> security, it is not by itself sufficient to build secure systems. Secure
> coding, static analysis, hardware tamper detection, JTAG locking, and many
> more techniques should be implemented as well.
> Code signing also does nothing to protect against reverse engineering. It is a
> distinct technique from firmware encryption.

> ### ECDSA

Reviewer: I think this section is quite dry, and less exciting than FIrmware Signing implementation. 

Maybe you can quickly say `We use ECDSA, which we talk about [later in the post](#ecdsa)` and jump straight to the example, and place this section later.

Author: I added a "jump to implementation" at the top of the theory section. At the same time, I think this needs to be before the implementation, because it details the use of SHA-256 which is key to the implementation.

---
title: "Authenticate firmware builds with code signing"
description: WIP
author: francois
---

<!-- excerpt start -->
<!-- excerpt end -->

{% include newsletter.html %}
{% include toc.html %}

## Firmware Signing Explained

> ### What is code signing

Reviewer: You can probably remove this specific header

Author: :+1:

Code signing is a method of proving a file was created by a trusted source and
has not been tampered with. This is achieved by creating a signature for the
file: a token which can be verified but not forged.

We all use signatures in our day to day life, by scribbling our name at the
bottom of our credit card receipts, contracts, and bank checks. These signatures
can be verified quite easily by comparing them to previous signatures, but are
hard to forge because they rely on an individual's muscle memory.

Code signing uses cryptographic algorithms to achieve a similar goal. The
resulting signatures can only be generated by people who know a given password,
also known as a secret key, but can be verified as genuine by anyone.

Additionally, the signatures are tied to a given firmware binary so should the
binary change by so much as 1 bit the signature would no longer be valid.

### Why sign our firmware

By implementing signature verification in our bootloader we can identify
whether or not a given firmware update was provided by the manufacturer, or if

> it has been tempered with. The bootloader can then decide to either warn the

Reviewer: ```suggestion
it has been tampered with. The bootloader can then decide to either warn the
```

user, void the device's warranty, or simply refuse to run the unauthenticated
binary.

With more and more devices connected to the internet, security is an
increasingly hot topic in firmware development. A device which accepts firmware
updates over the wireless or internet connectivity but does not verify it opens

> itself to compromise. By feeding it with a malicious firmware image, attacker

Reviewer: ```suggestion
itself to compromise. By feeding it with a malicious firmware image, an attacker
```

might:
* Brick the device, or the whole fleet
* Snoop on end users and compromise their privacy and security
* Strategically malfunction at a critical time

> These are highly undesirable outcome, which can be effected at scale due to the

Reviewer: ```suggestion
These are highly undesirable outcomes, which can be effected at scale due to the
```

internet of things. In 2020, it is reckless to implement firmware update for our
systems without some form of authentication.

> **What signing is not**: code signing is an important component of firmware
> security, it is not by itself sufficient to build secure systems. Secure
> coding, static analysis, hardware tamper detection, JTAG locking, and many
> more techniques should be implemented as well.
> Code signing also does nothing to protect against reverse engineering. It is a
> distinct technique from firmware encryption.

### ECDSA

Several algorithms can be used to sign firmware, including RSA, DSA, and ECDSA.

> In this post, we focus on ECDSA for a few reason:

Reviewer: ```suggestion
In this post, we focus on ECDSA for a few reasons:
```

micro-ecc.
4. Small Footprint: ECDSA implementations are very small (single digit kB), and
   require smaller keys than RSA or DSA for similar levels of security. This
saves both code space and RAM and makes ECDSA well suited to embedded
environments.

Understanding the math behind ECDSA is outside of the scope of this article, but
here is a high level overview of how the process works: 
1. A cryptographic hash of the firmware  binary is created. Any cryptographic
   hashing algorithm should work, though SHA-2 family hashes are recommended. In
the case of SHA-256, this yields a 32-byte number.
2. A signature is generated using a private key and the cryptographic hash.
   This signature is distributed alongside the firmware and a public key. The
signature may not be deterministic, so don't fret if multiple invocations of
your ECDSA code yield different signatures. This signature is a pair of
integers, each 32 bytes long.
3. To verify the binary, a SHA-256 hash is once again computed for our firmware
   binary.
4. The public key and the hash can be used to verify the signature was generated
   using matching inputs.

## Firmware Signing Implementation

### Setup

> Like we did in previous post, we use [Renode]({% post_url
> 2020-03-23-intro-to-renode %}) to run the examples in this post. Our previous
> [firmware update post]({% post_url 2020-06-23-device-firmware-update-cookbook
> %}) contains detailed instructions. In short:

Reviewer: ```suggestion
Like we did our previous post, the [Firmware Update Cookbook]({% post_url 2020-06-23-device-firmware-update-cookbook%}), we use [Renode]({% post_url
2020-03-23-intro-to-renode %}) to run the examples in this post. The previous post contains detailed instructions, but in short:
```

```
# Clone the repository & navigate to the example
$ git clone https://github.com/memfault/interrupt.git
$ cd examples/fwup-signing

# Build project and start Renode
$ make && ./start.sh
```

One change we did make from previous posts is run in headless mode. This is done
by invoking Renode with the following flags:

```
mono64 $RENODE_EXE_PATH renode-config.resc --port 4444 --disable-xwt
```

> We can then use telnet to connect to the renode console:

Reviewer: ```suggestion
We can then use telnet to connect to the Renode console:
```

    span_width = 100;
    // Set labels to nodes.
    C [label = "Bootloader"];
    A [label = "App Loader"];
    B [label = "Application"];
    C -> A [label = "Loads", fontsize=8];
    A -> B [label = "Ld, Updt", fontsize=8];

    E [label = "Updater"];
    A -> E [label = "Ld, Updt", fontsize=8];
    E -> A [label = "Updates", fontsize=8];
    C -> E [label = "Loads", style=dashed, fontsize=8];

    group {
        label = "Slot 0";
        color = "PaleGreen";
        C;
    }
    group {
        label = "Slot 1";
        color = "LightPink";
        A;
    }
    group {
        label = "Slot 2";
        color = "LemonChiffon";
        B; E;
    }
}
{% endblockdiag %}{:.diag4}

With the following functionality for each block:

1. Bootloader: a simple program whose sole job is to load the Loader, and fallback
   to another image if the Loader is invalid.
2. Loader: a program that can verify our Application image, load it, and update
   it.
3. Application: our main code, which does not do any updates itself
4. Updater: a program that temporarily replaces the Application and can update
   the Loader.

The Loader, the Application, and the Updater all need to be signed since they

> can be updated. Components that updates other components need to implement our

Reviewer: ```suggestion
can be updated. Components that update other components need to implement our
```

signature verification algorithm, in our case that's the Loader and the Updater.

### Generating public/private key pairs

> Before we can begin, we must generate a pair of keys. Our **private key** also
> known as **signing key** is used to create the signatures, and should be kept

Reviewer: ```suggestion
Before we can begin, we must generate a pair of keys. Our **private key**, also
known as a **signing key**, is used to create the signatures, and should be kept
```

private. Anyone with access to the key will be able to sign firmware on your

> behalf. Our **public key** also known as **validation key** is used to verify
> signature. It can be freely distributed.

Reviewer: ```suggestion
behalf. Our **public key**, also known as a **validation key**, is used to verify
the signature. It can be freely distributed.
```

Several tools can be used to generate our key pair but the simplest is

> `openssl`, a cross platform cryptography tools.

Reviewer: ```suggestion
`openssl`, a cross platform cryptography toolset.
```

+        # Seek to beginning of signature
+        f.seek(32)
+        # Write the signature in place
+        f.write(signature)

 if __name__ == "__main__":
@@ -54,6 +73,7 @@
         description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter,
     )
     parser.add_argument("bin", action="store")
+    parser.add_argument("pk", action="store")
     args = parser.parse_args()

-    patch_binary_payload(args.bin)
+    patch_binary_payload(args.bin, args.pk)
```

Since the script now takes an argument, we must modify our Makefile as well

```diff
-$(BUILD_DIR)/$(PROJECT)-%.bin: $(BUILD_DIR)/$(PROJECT)-%.elf
+$(BUILD_DIR)/$(PROJECT)-%.bin: $(BUILD_DIR)/$(PROJECT)-%.elf $(PK_PEM_PATH)
        $(ECHO) "  OBJCOPY      $@"
-       $(Q)$(OCPY) $^ $@ -O binary
+       $(Q)$(OCPY) $< $@ -O binary
        $(ECHO) "  PATCH_IMAGE  $@"
-       $(Q)$(PYTHON) patch_image_header.py $@ > /dev/null
+       $(Q)$(PYTHON) patch_image_header.py $@ $(PK_PEM_PATH) > /dev/null
```

Where `PK_PEM_PATH` is the path to your private key.<!-- TODO: Add a note about
security -->

We now have a build process that generates signed binaries!

### Verifying signatures in our Loader

> Now that our firmware builds are signed, lets verify those signatures in our

Reviewer: ```suggestion
Now that our firmware builds are signed, let's verify those signatures in our
```

    7c:4f:d1:7e:a5
ASN1 OID: secp256k1
```

We take the `pub` section, remove the leading `04` bytes, and translate it to a
C array:

```c
static const uint8_t PUBKEY[] = {
  0xd0, 0xe6, 0xa7, 0xa5, 0x4e, 0x33, 0x0e, 0xbb, 0xd9, 0x9e, 0xe6, 0x8f, 0x59,
  0xff, 0xb6, 0xc1, 0x19, 0x76, 0x28, 0x60, 0x88, 0x16, 0x6a, 0x17, 0x8b, 0x7b,
  0xe0, 0x66, 0xcf, 0x7b, 0x71, 0x0d, 0xf5, 0xcc, 0x95, 0x76, 0x22, 0xae, 0x0e,
  0xa4, 0xef, 0x49, 0xbd, 0x07, 0x2a, 0x71, 0x49, 0x84, 0x49, 0x78, 0xeb, 0x34,
  0xe5, 0x78, 0xb3, 0xa7, 0x96, 0x48, 0x89, 0x7c, 0x4f, 0xd1, 0x7e, 0xa5
  // 64 bytes
};
```

We can then verify that uECC is able to read the key correctly by calling
`uECC_valid_public_key`. We add the following code somewhere in our `main`
function at boot:

```c
#include <micro-ecc/uECC.h>

const struct uECC_Curve_t *curve = uECC_secp256k1();
if (!uECC_valid_public_key(PUBKEY, curve)) {
    printf("Public key is NOT valid\n");
} else {
    printf("Public key is valid\n");
}
```

Note that we had to select the same curve we used to generate the key, here
`secp2561k`.

Last but not least, we need to verify the signature itself. This requires
computing the SHA-256, and using `uECC_verify`.

> Many MCUs have hardware implementations of SHA-256, but in our case we used a

Reviewer: ```suggestion
Many MCUs have hardware implementations of SHA-256, but in our case, we used a
```

software implementation provided by a library called CIFRA[^cifra]. We only need
to add two files from CIFRA to our build: `cifra/src/sha256.c` and `cifra/src/blockwise.c`,
as well as two paths to our include path: `cifra/src` and `cifra/src/ext`.

With that, I implemented a simple sha256 wrapper function which computes the
hash of a given buffer:

```c
static void prv_sha256(const void *buf, uint32_t size, uint8_t *hash_out)
{
  cf_sha256_context ctx;
  cf_sha256_init(&ctx);
  cf_sha256_update(&ctx, buf, size);
  cf_sha256_digest_final(&ctx, hash_out);
}
```

> Note that `cf_sha256_update` can be called mutliple time, so if you are

Reviewer: ```suggestion
Note that `cf_sha256_update` can be called mutliple times, so if you are
```
